packages/@n8n/ai-workflow-builder.ee/src/subgraphs/configurator.subgraph.ts

```diff
@@ -0,0 +1,347 @@
+	ConfiguratorSubgraphConfig,
+	typeof ConfiguratorSubgraphState.State,
+	typeof ParentGraphState.State
+> {
+	name = 'configurator_subgraph';
+	description = 'Configures node parameters after structure is built';
+
+	private agent!: Runnable;
+	private toolMap!: Map<string, StructuredTool>;
+	private instanceUrl: string = '';
+
+	create(config: ConfiguratorSubgraphConfig) {
+		this.instanceUrl = config.instanceUrl ?? '';
+		// Create tools
+		const tools = [
+			createUpdateNodeParametersTool(
+				config.parsedNodeTypes,
+				config.llm, // Uses same LLM for parameter updater chain
+				config.logger,
+				config.instanceUrl,
+			),
+			createGetNodeParameterTool(),
+			createValidateConfigurationTool(config.parsedNodeTypes),
+		];
+		this.toolMap = new Map<string, StructuredTool>(tools.map((bt) => [bt.tool.name, bt.tool]));
+		// Create agent with tools bound
+		const systemPromptTemplate = ChatPromptTemplate.fromMessages([
+			[
+				'system',
+				[
+					{
+						type: 'text',
+						text: CONFIGURATOR_PROMPT,
+					},
+					{
+						type: 'text',
+						text: INSTANCE_URL_PROMPT,
+						cache_control: { type: 'ephemeral' },
+					},
+				],
+			],
+			['placeholder', '{messages}'],
+		]);
+
+		if (typeof config.llm.bindTools !== 'function') {
+			throw new LLMServiceError('LLM does not support tools', {
+				llmModel: config.llm._llmType(),
+			});
+		}
+
+		this.agent = systemPromptTemplate.pipe(config.llm.bindTools(tools.map((bt) => bt.tool)));
+
+		/**
+		 * Agent node - calls configurator agent
+		 * Context is already in messages from transformInput
+		 */
+		const callAgent = async (state: typeof ConfiguratorSubgraphState.State) => {
+			// Apply cache markers to accumulated messages (for tool loop iterations)
+			applySubgraphCacheMarkers(state.messages);
+
+			// Messages already contain context from transformInput
+			const response = (await this.agent.invoke({
+				messages: state.messages,
+				instanceUrl: state.instanceUrl ?? '',
+			})) as BaseMessage;
+
+			return { messages: [response] };
+		};
+
+		// Build the subgraph
+		const subgraph = new StateGraph(ConfiguratorSubgraphState)
+			.addNode('agent', callAgent)
+			.addNode('tools', async (state) => await executeSubgraphTools(state, this.toolMap))
+			.addNode('process_operations', processOperations)
+			.addEdge('__start__', 'agent')
+			// Map 'tools' to tools node, END is handled automatically
+			.addConditionalEdges('agent', createStandardShouldContinue(MAX_CONFIGURATOR_ITERATIONS))
+			.addEdge('tools', 'process_operations')
+			.addEdge('process_operations', 'agent'); // Loop back
+
+		return subgraph.compile();
+	}
+
+	transformInput(parentState: typeof ParentGraphState.State) {
+		const userRequest = extractUserRequest(parentState.messages);
+
+		// Build context parts for Configurator
+		const contextParts: string[] = [];
+
+		// 1. User request (primary)
+		if (userRequest) {
+			contextParts.push('=== USER REQUEST ===');
```

Comment:
Out of interest, do you think for the transform input/output it would be easier to read if we did something like:
```ts
const context = `=== USER REQUEST ===
${userRequest}
${parentState.discoveryContext?.bestPractices ?? ''}
=== WORKFLOW TO CONFIGURE ===
${buildWorkflowJsonBlock(parentState.workflowJSON)}
`.split('\n');
```
And so on - I feel like this might be a little easier to comprehend/adjust at a glance.